clang-tools-extra/include-cleaner/tool/IncludeCleaner.cpp

```diff
@@ -305,8 +307,84 @@ int main(int argc, const char **argv) {
     }
   }
 
-  clang::tooling::ClangTool Tool(OptionsParser->getCompilations(),
-                                 OptionsParser->getSourcePathList());
+  auto VFS = llvm::vfs::getRealFileSystem();
+  auto &CDB = OptionsParser->getCompilations();
+  // CDBToAbsPaths is a map from the path in the compilation database to the
+  // writable absolute path of the file.
+  std::map<std::string, std::string> CDBToAbsPaths;
+  if (Edit) {
+    // if Edit is enabled, `Factory.editedFiles()` will contain the final code,
+    // along with the path given in the compilation database. That path can be
+    // absolute or relative, and if it is relative, it is relative to the
+    // "Directory" field in the compilation database. We need to make it
+    // absolute to write the final code to the correct path.
+    // There are several cases to consider:
+    // 1. The "Directory" field isn't same as the current working directory.
+    // 2. The file path resolved from the "Directory" field is not writable.
+    // For these cases, we need to find a writable path for the file.
+    // To effectively handle these cases, we only need to consider
+    // the files from `getSourcePathList()` that are present in the compilation
+    // database.
+    for (auto &Source : OptionsParser->getSourcePathList()) {
+      llvm::SmallString<256> AbsPath(Source);
+      if (auto Err = VFS->makeAbsolute(AbsPath)) {
+        llvm::errs() << "Failed to get absolute path for " << Source << " : "
+                     << Err.message() << '\n';
+        return 1;
+      }
+      std::vector<clang::tooling::CompileCommand> Cmds =
+          CDB.getCompileCommands(AbsPath);
+      if (Cmds.empty()) {
+        // Try with the original path.
+        Cmds = CDB.getCompileCommands(Source);
+        if (Cmds.empty()) {
+          continue;
+        }
+      }
+      // We only need the first command to get the directory.
+      auto Cmd = Cmds[0];
+      llvm::SmallString<256> CDBPath(Cmd.Filename);
+      std::string Directory(Cmd.Directory);
+
+      if (llvm::sys::path::is_absolute(CDBPath)) {
+        // If the path in the compilation database is already absolute, we don't
+        // need to do anything.
+        CDBToAbsPaths[static_cast<std::string>(CDBPath)] =
+            static_cast<std::string>(AbsPath);
+      } else {
+        auto Sept = llvm::sys::path::get_separator();
+        // First, try to find the file based on the compilation database.
+        llvm::Twine FilePathTwine = Directory + Sept + CDBPath;
+        llvm::SmallString<256> FilePath;
+        FilePathTwine.toVector(FilePath);
+        // Check if it is writable.
+        if (llvm::sys::fs::access(FilePath, llvm::sys::fs::AccessMode::Write) !=
+            std::error_code()) {
```

Comment:
i don't see why we need to complicate the logic by checking for this. is there any reasons to not always map to file-path relative to current process?

i am pretty convinced that the user intention is to almost always edit the file path as referred to during invocation, e.g. `cd /foo && clang-include-cleaner path/to/foo.cc` is always meant to edit `/foo/path/to/foo.cc` no matter what file path we get from CDB.

i can see how this might also work, but I prefer to maintain code that has as few special cases as possible. so unless something is indeed breaking with this simplification, can you please get rid of this case?

Reply:
It turns out we don't need to check for this. I removed those checks.